test/test_datapipe.py

```diff
@@ -589,19 +598,40 @@ def test_serializable(self):
         dp_skip_comparison = {dp.iter.FileLister, dp.iter.FileOpener, dp.iter.StreamReader, dp.iter.Shuffler}
         # These DataPipes produce multiple DataPipes as outputs and those should be compared
         dp_compare_children = {dp.iter.Demultiplexer, dp.iter.Forker}
+
         for dpipe, dp_args, dp_kwargs in picklable_datapipes:
             datapipe = dpipe(input_dp, *dp_args, **dp_kwargs)  # type: ignore[call-arg]
-            serialized_dp = pickle.dumps(datapipe)
-            deserialized_dp = pickle.loads(serialized_dp)
+            serialized_dp = pickle.dumps(datapipe)  # Making sure dpipe is picklable
+            _ = pickle.loads(serialized_dp)  # Making sure dpipe is loadable
             if dpipe not in dp_skip_comparison:
-                if dpipe in dp_compare_children:
-                    for c1, c2 in zip(list(datapipe), list(deserialized_dp)):
-                        self.assertEqual(list(c1), list(c2))
-                else:
-                    self.assertEqual(list(datapipe), list(deserialized_dp))
+                # 1. Testing for serialization before any iteration starts
+                self._serialization_test_helper(datapipe, dpipe in dp_compare_children)
+
+                # 2. Testing for serialization after DataPipe is partially read
+                if dpipe not in dp_compare_children:
+                    it = iter(datapipe)
+                    _ = next(it)
+                    self._serialization_test_helper(datapipe)
+
+                    # 3. Testing for serialization after DataPipe is fully read
+                    _ = list(it)
+                    self._serialization_test_helper(datapipe)
+                else:  # DataPipes that need to compare children
+                    it1, it2 = iter(datapipe[0]), iter(datapipe[1])
+                    _, _ = next(it1), next(it2)
+                    self._serialization_test_helper(datapipe, has_two_children=True)
```

Comment:
You can drop has_two_children and just call _serialization_test_helper twice for dp1 and dp2